File: web/semantics.js
function isBetaRedex(expr) {
    return expr instanceof Application && expr.left instanceof Abstraction;
}

function betaReduceOnce(expr) {
    if (expr instanceof Variable) {
        return [expr, false];
    }
    
    else if (expr instanceof Abstraction) {
        const [bodyReduced, wasReduced] = betaReduceOnce(expr.body);
        if (wasReduced) {
            return [new Abstraction(expr.param, bodyReduced), true];
        }
        return [expr, false];
    }
    
    else if (expr instanceof Application) {
        if (isBetaRedex(expr)) {
            const abstraction = expr.left;
            const argument = expr.right;
            return [abstraction.body.substitute(abstraction.param, argument), true];
        }
        
        const [leftReduced, leftWasReduced] = betaReduceOnce(expr.left);
        if (leftWasReduced) {
            return [new Application(leftReduced, expr.right), true];
        }
        
        const [rightReduced, rightWasReduced] = betaReduceOnce(expr.right);
        if (rightWasReduced) {
            return [new Application(expr.left, rightReduced), true];
        }
        
        return [expr, false];
    }
    
    throw new TypeError(`Unknown expression type: ${expr.constructor.name}`);
}

function betaReduceNormalOrder(expr, maxSteps = 1000) {
    let steps = 0;
    let currentExpr = expr;
    let reductionSteps = [];
    
    while (steps < maxSteps) {
        const [reducedExpr, wasReduced] = betaReduceOnce(currentExpr);
        
        if (!wasReduced) {
            return [currentExpr, steps, true, reductionSteps];
        }
        
        reductionSteps.push(reducedExpr.toString());
        currentExpr = reducedExpr;
        steps++;
    }
    
    return [currentExpr, steps, false, reductionSteps];
}
function betaReduceApplicativeOrder(expr, maxSteps = 1000) {
    function findInnermostRedex(expr) {
        if (expr instanceof Variable) {
            return [null, null];
        }
        
        else if (expr instanceof Abstraction) {
            const [path, foundExpr] = findInnermostRedex(expr.body);
            if (path !== null) {
                return [['body', ...path], foundExpr];
            }
            return [null, null];
        }
        
        else if (expr instanceof Application) {
            const [rightPath, rightFoundExpr] = findInnermostRedex(expr.right);
            if (rightPath !== null) {
                return [['right', ...rightPath], rightFoundExpr];
            }
            
            const [leftPath, leftFoundExpr] = findInnermostRedex(expr.left);
            if (leftPath !== null) {
                return [['left', ...leftPath], leftFoundExpr];
            }
            
            if (isBetaRedex(expr)) {
                return [[], expr];
            }
            
            return [null, null];
        }
    }
    
    function replaceAtPath(expr, path, replacement) {
        if (path.length === 0) {
            return replacement;
        }
        
        if (expr instanceof Abstraction && path[0] === 'body') {
            return new Abstraction(
                expr.param,
                replaceAtPath(expr.body, path.slice(1), replacement)
            );
        }
        
        if (expr instanceof Application) {
            if (path[0] === 'left') {
                return new Application(
                    replaceAtPath(expr.left, path.slice(1), replacement),
                    expr.right
                );
            } else if (path[0] === 'right') {
                return new Application(
                    expr.left,
                    replaceAtPath(expr.right, path.slice(1), replacement)
                );
            }
        }
        
        throw new Error(`Invalid path ${path} for expression ${expr}`);
    }
    
    let steps = 0;
    let currentExpr = expr;
    let reductionSteps = [];
    
    while (steps < maxSteps) {
        const [path, redex] = findInnermostRedex(currentExpr);
        
        if (redex === null) {
            return [currentExpr, steps, true, reductionSteps];
        }
        
        const abstraction = redex.left;
        const argument = redex.right;
        const reducedRedex = abstraction.body.substitute(abstraction.param, argument);
        
        currentExpr = replaceAtPath(currentExpr, path, reducedRedex);
        reductionSteps.push(currentExpr.toString());
        steps++;
    }
    
    return [currentExpr, steps, false, reductionSteps];
}

function isNormalForm(expr) {
    if (expr instanceof Variable) {
        return true;
    }
    
    else if (expr instanceof Abstraction) {
        return isNormalForm(expr.body);
    }
    
    else if (expr instanceof Application) {
        if (expr.left instanceof Abstraction) {
            return false;
        }
        
        return isNormalForm(expr.left) && isNormalForm(expr.right);
    }
    
    throw new TypeError(`Unknown expression type: ${expr.constructor.name}`);
}

function reduceToNormalForm(expr, strategy = "normal", maxSteps = 1000) {
    if (strategy === "normal") {
        return betaReduceNormalOrder(expr, maxSteps);
    } else if (strategy === "applicative") {
        return betaReduceApplicativeOrder(expr, maxSteps);
    } else {
        throw new Error(`Unknown evaluation strategy: ${strategy}`);
    }
}

File: lambda_calculus/semantics.py
from lambda_calculus.syntax import Expr, Variable, Abstraction, Application


def is_beta_redex(expr):
    return isinstance(expr, Application) and isinstance(expr.left, Abstraction)


def beta_reduce_once(expr):
    if isinstance(expr, Variable):
        return expr, False
    
    elif isinstance(expr, Abstraction):
        body_reduced, was_reduced = beta_reduce_once(expr.body)
        if was_reduced:
            return Abstraction(expr.param, body_reduced), True
        return expr, False
    
    elif isinstance(expr, Application):
        if is_beta_redex(expr):
            abstraction = expr.left
            argument = expr.right
            return abstraction.body.substitute(abstraction.param, argument), True
        
        left_reduced, was_reduced = beta_reduce_once(expr.left)
        if was_reduced:
            return Application(left_reduced, expr.right), True
        
        right_reduced, was_reduced = beta_reduce_once(expr.right)
        if was_reduced:
            return Application(expr.left, right_reduced), True
        
        return expr, False
    
    raise TypeError(f"Unknown expression type: {type(expr)}")


def beta_reduce_normal_order(expr, max_steps=1000):
    steps = 0
    current_expr = expr
    
    while steps < max_steps:
        reduced_expr, was_reduced = beta_reduce_once(current_expr)
        if not was_reduced:
            return current_expr, steps, True
        
        current_expr = reduced_expr
        steps += 1
    return current_expr, steps, False


def beta_reduce_applicative_order(expr, max_steps=1000):
    def find_innermost_redex(expr):
        if isinstance(expr, Variable):
            return None, None
        
        elif isinstance(expr, Abstraction):
            path, found_expr = find_innermost_redex(expr.body)
            if path is not None:
                return ['body'] + path, found_expr
            return None, None
        
        elif isinstance(expr, Application):
            path, found_expr = find_innermost_redex(expr.right)
            if path is not None:
                return ['right'] + path, found_expr
            
            path, found_expr = find_innermost_redex(expr.left)
            if path is not None:
                return ['left'] + path, found_expr
            
            if is_beta_redex(expr):
                return [], expr
            
            return None, None
    
    def replace_at_path(expr, path, replacement):
        if not path:
            return replacement
        
        if isinstance(expr, Abstraction) and path[0] == 'body':
            return Abstraction(expr.param, replace_at_path(expr.body, path[1:], replacement))
        
        if isinstance(expr, Application):
            if path[0] == 'left':
                return Application(replace_at_path(expr.left, path[1:], replacement), expr.right)
            elif path[0] == 'right':
                return Application(expr.left, replace_at_path(expr.right, path[1:], replacement))
        
        raise ValueError(f"Invalid path {path} for expression {expr}")
    
    steps = 0
    current_expr = expr
    
    while steps < max_steps:
        path, redex = find_innermost_redex(current_expr)
        if redex is None:
            return current_expr, steps, True
        
        abstraction = redex.left
        argument = redex.right
        reduced_redex = abstraction.body.substitute(abstraction.param, argument)
        
        current_expr = replace_at_path(current_expr, path, reduced_redex)
        steps += 1
    
    return current_expr, steps, False


def is_normal_form(expr):
    if isinstance(expr, Variable):
        return True
    
    elif isinstance(expr, Abstraction):
        return is_normal_form(expr.body)
    
    elif isinstance(expr, Application):
        if isinstance(expr.left, Abstraction):
            return False
        
        return is_normal_form(expr.left) and is_normal_form(expr.right)
    
    raise TypeError(f"Unknown expression type: {type(expr)}")


def reduce_to_normal_form(expr, strategy="normal", max_steps=1000):
    if strategy == "normal":
        return beta_reduce_normal_order(expr, max_steps)
    elif strategy == "applicative":
        return beta_reduce_applicative_order(expr, max_steps)
    else:
        raise ValueError(f"Unknown evaluation strategy: {strategy}")
